New Blog: Enabling branch deployments through IssueOps with GitHub Actions (#93)
yuhattor/hattori.dev: content/githubblog/2023-02-02-enabling-branch-deployments-through-issueops-with-github-actions.md

---

> title: "GitHub ActionsでIssueOpsによるブランチデプロイメントを可能にする。"

Reviewer: ```suggestion
title: "GitHub ActionsでIssueOpsによるブランチデプロイメントを可能にする"
```

> subtitle: "もし開発者がブランチデプロイを活用したいが、ChatOpsのフルスタックをリポジトリに統合していない場合はどうすればいいでしょうか？私たちは、すべての開発者がGitHubリポジトリから簡単にブランチデプロイを利用できる方法を見つけたいと思い、branch-deploy Actionが生まれました!"

Reviewer: ```suggestion
subtitle: "開発者がブランチデプロイを活用したいのに、ChatOpsのフルスタックをリポジトリに統合していない場合はどうすればいいでしょうか？ すべての開発者がGitHubリポジトリから簡単にブランチデプロイを利用できる方法を見つけるために、branch-deploy Action が生まれました!"
```

englishtitle: "Enabling branch deployments through IssueOps with GitHub Actions"
date: "2023-02-02"
cardurl: "https://github.blog/2023-02-02-enabling-branch-deployments-through-issueops-with-github-actions/"
author: "Grant Birkinbine"
description: "At GitHub, the branch deploy model is ubiquitous and it is the standard way we ship code to production, and it has been for years. We released details about how we perform branch deployments with ChatOps all the way back in 2015 ."
coverimage: "https://github.blog/wp-content/uploads/2022/09/Open-Source-Engineering@2x.png?resize=1600%2C850"
category: "Engineering,Open Source,automation,Deployment,GitHub Actions"
---

<p>GitHubでは、ブランチ・デプロイ・モデルはどこにでもあり、私たちがコードをプロダクションに出荷する標準的な方法であり、それは何年も前からそうでした。私たちは、<a href="https://github.blog/2015-06-02-deploying-branches-to-github-com/">2015</a>年にChatOpsでブランチデプロイを行う方法の詳細を公開しました。</p>

> <p>私たちはほとんどのリポジトリでChatOpsを使用してブランチデプロイメントを実行することができますが、ChatOpsが単にうまくいかない状況もいくつかあります。もし開発者がブランチデプロイを活用したいが、リポジトリにChatOpsのフルスタックを統合していない場合はどうすればいいのでしょうか？私たちは、すべての開発者がGitHubのリポジトリから簡単にブランチ・デプロイを利用できる方法を見つけたいと思い、branch-deploy Actionが生まれました!</p>

Reviewer: ```suggestion
<p>私たちはほとんどのリポジトリでChatOpsを使用してブランチデプロイメントを実行することができますが、ChatOpsが単にうまくいかない状況もいくつかあります。もし開発者がブランチデプロイを活用したいが、リポジトリにChatOpsのフルスタックを統合していない場合はどうすればいいのでしょうか？私たちは、すべての開発者がGitHubのリポジトリから簡単にブランチ・デプロイを利用できる方法を見つけるために、branch-deploy Action を作りました！</p>
```

<p><img decoding="async" src="https://github.blog/wp-content/uploads/2023/02/usage.gif" alt="Gif demonstrating how to us the branch-deploy Action."/></p>
<h2 id="how-does-github-use-this-action">GitHubはこのアクションをどのように使っているのか？<a href="#how-does-github-use-this-action" class="heading-link pl-2 text-italic text-bold" aria-label="How Does GitHub use this Action?"></a></h2>

> <p>GitHubは主にChatOpsと<a href="https://hubot.github.com/">Hubot</a>)を使って、可能な限りブランチデプロイを促進するようにしています。ChatOpsが使えない場合は、このbranch-deploy Actionで代用しています。私たちのユースケースの大半は、インフラストラクチャー・アズ・コード（IaC）リポジトリで、インフラの変更をTerraformを使ってデプロイしています。GitHubはこのActionを多くの内部リポジトリで使っていますし、<a href="https://www.npmjs.com/">npmも</a>そうです。他にも多くの公共、オープンソース、企業組織がこのActionを採用し、コードを本番環境に出荷するのに役立っています。</p>

Reviewer: ```suggestion
<p>GitHubは主にChatOpsと<a href="https://hubot.github.com/">Hubot</a>)を使って、可能な限りブランチデプロイを促進するようにしています。ChatOpsが使えない場合は、このbranch-deploy Actionで代用しています。私たちのユースケースの大半は、Infrastructure as Code (IaC)リポジトリで、インフラの変更をTerraformを使ってデプロイしています。GitHubはこのActionを多くの内部リポジトリで使っていますし、<a href="https://www.npmjs.com/">npmも</a>そうです。他にも多くの公共、オープンソース、企業組織がこのActionを採用し、コードを本番環境に出荷するのに役立っています。</p>
```

<h2 id="understanding-the-branch-deploy-model">ブランチ・デプロイ・モデルを理解する<a href="#understanding-the-branch-deploy-model" class="heading-link pl-2 text-italic text-bold" aria-label="Understanding the branch deploy model"></a></h2>
<p>branch-deploy アクションを理解する前に、まずブランチデプロイモデルとは何か、そしてなぜそれが有用なのかを理解しましょう。</p>
<p>ブランチデプロイモデルを理解するために、まず伝統的な<strong>デプロイ → マージ</strong>モデルを見てみましょう。それは次のようなものです。</p>
<ol>

> <li>ブランチを作成する。</li>

Reviewer: ```suggestion
<li>ブランチを作成する</li>
```

> <li>ブランチにコミットを追加する。</li>

Reviewer: ```suggestion
<li>ブランチにコミットを追加する</li>
```

> <li>プルリクエストを発行する。</li>

Reviewer: ```suggestion
<li>プルリクエストを発行する。</li>
```
```suggestion
<li>プルリクエストを発行する</li>
```

<p>では、<strong>ブランチのデプロイモデルを見て</strong>いきましょう。</p>
<ol>
<li>ブランチを作成する。</li>
<li>ブランチにコミットを追加する。</li>
<li>プルリクエストを作成する。</li>
<li>フィードバックとピアレビューを収集する</li>
<li>変更をデプロイする</li>
<li>検証する</li>
<li>あなたのブランチを<code>main</code>/<code>master</code>ブランチにマージする。 </li>
</ol>
<div class="image-frame image-frame-full border rounded-2 overflow-hidden d-flex flex-row flex-justify-center" style="background: #EAEEF2"><br />
<img decoding="async" loading="lazy" src="https://github.blog/wp-content/uploads/2023/02/branch-deploy.jpg?w=720&#038;resize=720%2C244" alt="Diagram outlining the steps of the branch deploy model, enumerated in the list above." width="720" height="244" class="aligncenter size-large wp-image-69855 width-fit" srcset="https://github.blog/wp-content/uploads/2023/02/branch-deploy.jpg?w=720&#038;resize=720%2C244 720w, https://github.blog/wp-content/uploads/2023/02/branch-deploy.jpg?w=300 300w" sizes="(max-width: 720px) 100vw, 720px" data-recalc-dims="1" /><br /></div>

> <p><code>メイン</code>ブランチが本当に安定したブランチになることはないため、マージデプロイモデルは本質的にリスクが高いです。デプロイに失敗したり、ロールバックが必要になったりした場合は、もう一度すべてのプロセスをたどって変更をロールバックすることになります。しかし、ブランチデプロイモデルでは、<code>メイン</code>ブランチは常に「良い」状態であり、いつでもブランチデプロイからデプロイを元に戻すことができます。ブランチデプロイモデルでは、ブランチが正常にデプロイされ、検証された時点で初めて変更を<code>メイン</code>ブランチにマージします。</p>

Reviewer: ```suggestion
<p><code>main</code>ブランチが本当に安定したブランチになることはないため、マージデプロイモデルは本質的にリスクが高いです。デプロイに失敗したり、ロールバックが必要になったりした場合は、もう一度すべてのプロセスをたどって変更をロールバックすることになります。しかし、ブランチデプロイモデルでは、<code>main</code>ブランチは常に「良い」状態であり、いつでもブランチデプロイからデプロイを元に戻すことができます。ブランチデプロイモデルでは、ブランチが正常にデプロイされ、検証された時点で初めて変更を<code>main</code>ブランチにマージします。</p>
```

> <p><em>注：これは、<a href="https://docs.github.com/en/get-started/quickstart/github-flow">GitHubフローと</a>呼ばれることもあります。</em></p>

Reviewer: ```suggestion
<p><em>注：これは、<a href="https://docs.github.com/en/get-started/quickstart/github-flow">GitHub flowと</a>呼ばれることもあります。</em></p>
```

<h3 id="key-concepts">主要な概念<a href="#key-concepts" class="heading-link pl-2 text-italic text-bold" aria-label="Key concepts"></a></h3>
<p>ブランチ・デプロイ・モデルの主要な概念です。</p>
<ul>
<li><code>main</code>ブランチは、常に安定したデプロイ可能なブランチであるとみなされます。</li>

> <li>すべての変更は、<code>メイン</code>ブランチにマージされる前に本番環境にデプロイされます。</li>

Reviewer: ```suggestion
<li>すべての変更は、<code>main</code>ブランチにマージされる前に本番環境にデプロイされます。</li>
```

> <li>ブランチデプロイメントをロールバックするには、<code>メイン</code>ブランチをデプロイします。</li>

Reviewer: ```suggestion
<li>ブランチデプロイメントをロールバックするには、<code>main</code>ブランチをデプロイします。</li>
```

<p>noopデプロイが検出されたので、このアクションは<code>noop</code>変数を<code>trueに</code>出力します。IssueOpsコマンドを実行する正しい権限がある場合、このアクションは<code>continue</code>変数も<code>trueに</code>出力します。<code>fake noop deployという</code>名前のステップが実行され、<code>fake regular deployの</code>ステップはスキップされます。</p>
<h4 id="3-after-your-noop-deploy-completes-you-would-typically-run-deploy-to-execute-the-actual-deployment-fake-regular-deploy">3.3. noop deployが完了したら、通常、<code>.deployを</code>実行して、実際のデプロイメント、<code>偽の通常のデプロイメントを</code>実行することになります。<a href="#3-after-your-noop-deploy-completes-you-would-typically-run-deploy-to-execute-the-actual-deployment-fake-regular-deploy" class="heading-link pl-2 text-italic text-bold" aria-label="3. After your noop deploy completes, you would typically run &lt;code&gt;.deploy&lt;/code&gt; to execute the actual deployment, &lt;code&gt;fake regular deploy&lt;/code&gt;."></a></h4>
<h2 id="features">特徴<a href="#features" class="heading-link pl-2 text-italic text-bold" aria-label="Features"></a></h2>
<p>branch-deploy Action の最も優れた点は、あらゆるデプロイメントターゲットやユースケースに対して<strong>高度に</strong>カスタマイズ可能であることです。以下は、このActionに搭載されている機能の一部です。</p>
<ul>

> <li><img src="https://s.w.org/images/core/emoji/14.0.0/72x72/1f50d.png" alt="🔍" class="wp-smiley" style="height: 1em; max-height: 1em;" /> IssueOpsコマンドがプルリクエストで使用されると検出します。 </li>

Reviewer: ```suggestion
<li><img src="https://s.w.org/images/core/emoji/14.0.0/72x72/1f50d.png" alt="🔍" class="wp-smiley" style="height: 1em; max-height: 1em;" /> IssueOpsコマンドがプルリクエストで使用されると検出します </li>
```

> <li><img src="https://s.w.org/images/core/emoji/14.0.0/72x72/1f4dd.png" alt="📝" class="wp-smiley" style="height: 1em; max-height: 1em;" /> 設定可能：コマンドの構文、環境、noopトリガー、ベースブランチ、反応などを選択できます。</li>

Reviewer: ```suggestion
<li><img src="https://s.w.org/images/core/emoji/14.0.0/72x72/1f4dd.png" alt="📝" class="wp-smiley" style="height: 1em; max-height: 1em;" /> 設定可能：コマンドの構文、環境、noopトリガー、ベースブランチ、反応などを選択できます</li>
```

> <li><img src="https://s.w.org/images/core/emoji/14.0.0/72x72/2705.png" alt="✅" class="wp-smiley" style="height: 1em; max-height: 1em;" /> リポジトリに設定されたブランチ保護設定を尊重します。</li>

Reviewer: ```suggestion
<li><img src="https://s.w.org/images/core/emoji/14.0.0/72x72/2705.png" alt="✅" class="wp-smiley" style="height: 1em; max-height: 1em;" /> リポジトリに設定されたブランチ保護設定を尊重します</li>
```

> <li><img src="https://s.w.org/images/core/emoji/14.0.0/72x72/1f4ac.png" alt="💬" class="wp-smiley" style="height: 1em; max-height: 1em;" /> IssueOps コマンドにコメントし、反応します。</li>

Reviewer: ```suggestion
<li><img src="https://s.w.org/images/core/emoji/14.0.0/72x72/1f4ac.png" alt="💬" class="wp-smiley" style="height: 1em; max-height: 1em;" /> IssueOps コマンドにコメントし、反応します</li>
```

> <li><img src="https://s.w.org/images/core/emoji/14.0.0/72x72/1f680.png" alt="🚀" class="wp-smiley" style="height: 1em; max-height: 1em;" /> 簡単な設定でGitHubデプロイメントをトリガーします。</li>

Reviewer: ```suggestion
<li><img src="https://s.w.org/images/core/emoji/14.0.0/72x72/1f680.png" alt="🚀" class="wp-smiley" style="height: 1em; max-height: 1em;" /> 簡単な設定でGitHubデプロイメントをトリガーします</li>
```

> <li><img src="https://s.w.org/images/core/emoji/14.0.0/72x72/1f513.png" alt="🔓" class="wp-smiley" style="height: 1em; max-height: 1em;" /> デプロイロックにより、複数のデプロイの衝突を防止します。</li>

Reviewer: ```suggestion
<li><img src="https://s.w.org/images/core/emoji/14.0.0/72x72/1f513.png" alt="🔓" class="wp-smiley" style="height: 1em; max-height: 1em;" /> デプロイロックにより、複数のデプロイの衝突を防止します</li>
```
